File: MAY 23/12 May/2140. Solving Questions With Brainpower LC.cpp
class Solution {
public:
    long long mostPoints(vector<vector<int>>& questions) {
        int n = questions.size();
        vector<long long> dp(n+5,0);
        long long point, next;
        for(int i=n-1;i>=0;i--)
        {
            next = i + questions[i][1] + 1;
            point = questions[i][0];
            dp[i] = 0ll + point + dp[min(1ll*n,next)];
            dp[i] = max(dp[i], dp[i+1]);
        }
        return dp[0];
    }
};

File: APRIL 2025/01 April/2140. Solving Questions With Brainpower LC.java
class Solution {
    public long mostPoints(int[][] questions) {
        int N = questions.length;
        long[] dp = new long[N+1];
        int point, brainPower;
        long mostPoint = 0;
        for(int i=N-1; i>=0; i--) {
            point = questions[i][0];
            brainPower = questions[i][1];

            if(i+brainPower+1<N)
                dp[i] = Math.max(dp[i+1], point+dp[i+brainPower+1]);
            else
                dp[i] = Math.max(dp[i+1], point);
            // System.out.println(i+" "+dp[i]+" "+((i+brainPower<N)?dp[i+brainPower]:0));
            mostPoint = Math.max(mostPoint, dp[i]);
        }
        return dp[0];
    }
}
